File: Frontend/proyectoF/src/services/loan.service.js
import axios from 'axios';

const API_URL = 'http://localhost:8091/api/loans'; 

const LoanService = {
  // Related to Loan Simulation
  calculateMonthlyPayment: async (loanAmount, annualInterestRate, totalYears) => {
    try {
      const response = await axios.get(`${API_URL}/calculate`, {
        params: {
          loanAmount,
          annualInterestRate,
          totalYears
        }
      });
      return response.data; 
    } catch (error) {
      console.error('Error in simulate loan', error);
      throw error; 
    }
  },

  // Related to Loan Solicitude
  postLoanSolicitude: async (LoanSolicitude) => {
  try {
    const response = await axios.post(`${API_URL}/`, LoanSolicitude);
    console.log('Loan Solicitude Response:', response); 
    return response.data;

  } catch (error) {
    console.error('Error in post new loan solicitude in the database', error);
    throw error; 
  }
  },
  
  // Related to Loan Evaluation
  getPendingLoans: async () => {
    const response = await axios.get(`${API_URL}/pending`);
    return response.data;
  },

  updateLoan: async (loan) => {
    const response = await axios.put(`${API_URL}/`, loan);
    return response.data;
  },

  evaluateR1: async (quota, income) => {
    try {
        const response = await axios.post(`${API_URL}/evaluate/R1`, null, {
            params: {
                quota,
                income,
            },
        });
        return response.data; 
    } catch (error) {
        console.error('Error evaluating loan R1:', error);
        throw error;
    }
  },

  evaluateR3: async (yearsOfEmployment, isSelfEmployed) => {
    try {
        const response = await axios.post(`${API_URL}/evaluate/R3`, null, {
            params: {
                yearsOfEmployment,
                isSelfEmployed,
            },
        });
        return response.data;
    } catch (error) {
        console.error('Error evaluating loan R3:', error);
        throw error;
    }
  },

  evaluateR4: async (totalDebts, monthlyIncome) => {
    try {
      const response = await axios.post(`${API_URL}/evaluate/R4`, null, {
        params: {
          totalDebts,
          monthlyIncome,
        },
      });
      return response.data;
    } catch (error) {
      console.error('Error evaluating loan R4:', error);
      throw error;
    }
  },

  evaluateR5: async (loanAmount,propertyValue,propertyType) => {
    try {
      const response = await axios.post(`${API_URL}/evaluate/R5`, null, {
        params: {
          loanAmount,
          propertyValue,
          propertyType,
        },
      });
      return response.data;
    } catch (error) {
      console.error('Error evaluating loan R5:', error);
      throw error;
    }
  },

  evaluateR6: async (age, term) => {
    try {
      const response = await axios.post(`${API_URL}/evaluate/R6`, null, {
        params: {
          age,
          term,
        },
      });
      return response.data;
    } catch (error) {
      console.error('Error evaluating loan R6:', error);
      throw error;
    }
  },



}; 
export default LoanService;

File: Backend/v1.1/src/main/java/com/example/demo/services/LoanService.java
package com.example.demo.services;

import com.example.demo.entities.LoanEntity;
import com.example.demo.repositories.LoanRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class LoanService {
    @Autowired
    LoanRepository loanRepository;

    // Getters of loans
    public List<LoanEntity> getLoans(){ return (List<LoanEntity>) loanRepository.findAll();}
    public LoanEntity getLoanById(Long id){ return loanRepository.findById(id).get();}
    public List<LoanEntity> getLoanByRut(String rut){ return loanRepository.findByRut(rut);}

    // Creation
    public LoanEntity postLoanSolicitude(LoanEntity loanSolicitude){return loanRepository.save(loanSolicitude);}
    // Update
    public LoanEntity updateLoanSolicitude(LoanEntity loanSolicitude){ return loanRepository.save(loanSolicitude);}

    // P1 to calculate the monthly payment of the loan
    public int calculateMonthlyLoanPayment(int loanAmount, double annualInterestRate, int totalYears) {
        int months = totalYears * 12;
        double monthlyInterestRate = annualInterestRate / 12 / 100;
        double monthlyPayment = loanAmount * ((monthlyInterestRate * Math.pow(1 + monthlyInterestRate, months)) /
                (Math.pow(1 + monthlyInterestRate, months) - 1));
        return (int) Math.ceil(monthlyPayment);
    }

    // P4. For the solicitude revision
    public List<LoanEntity> getPendingLoans(){
        return loanRepository.findBySolicitudeStateNot("Approved");
    }

    // R1.
    public boolean R1cuoteIncomeRelation(int quota, int income) {
        // Verify the values of the quota and income
        if (quota <= 0 || income <= 0) {
            return false;
        }
        // Convertimos quota a float para obtener los decimales de la división
        float relation = (float) quota / income;
        return relation <= 0.35;
    }

    // R3
    public boolean R3evaluateEmploymentStability(int yearsOfEmployment, boolean isSelfEmployed) {
        if (isSelfEmployed) {
            return yearsOfEmployment >= 2; // Case of being self-employed
        } else {
            return yearsOfEmployment >= 1; // Case of being employed
        }
    }
    // R4
    public boolean R4ratioDebsIncome(int totalDebts, int monthlyIncome) {
        float ratio = (float) totalDebts / monthlyIncome;
        return !(ratio > 0.5);
    }

    public boolean R5maxAmount(int loanAmount, int propertyValue, int propertyType) {
        double maxAllowedLoan;
        switch (propertyType) {
            case 1:  // Primera Vivienda
                maxAllowedLoan = propertyValue * 0.8;
                break;
            case 2:  // Segunda Vivienda
                maxAllowedLoan = propertyValue * 0.7;
                break;
            case 3:  // Propiedades Comerciales
                maxAllowedLoan = propertyValue * 0.6;
                break;
            case 4:  // Remodelación
                maxAllowedLoan = propertyValue * 0.5;
                break;
            default:  // Tipo de propiedad no válido
                throw new IllegalArgumentException("Tipo de propiedad inválido");
        }
        return loanAmount <= maxAllowedLoan;
    }

    public boolean R6ageLimit(int age, int term) {
        return (term + age) <=  70;
    }



}
